[scroll-animations] Have named timeline code use Styleable

nmoucht opened the pull request:
#### f9880cf8ee2c4ddb9e6a5e7f91ff67a1ab34de31
<pre>
[scroll-animations] Have named timeline code use Styleable
<a href="https://bugs.webkit.org/show_bug.cgi?id=286781">https://bugs.webkit.org/show_bug.cgi?id=286781</a>
<a href="https://rdar.apple.com/143926206">rdar://143926206</a>

Reviewed by Antoine Quint.

Have named timeline code use Styleable.

* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::originatingElement):
(WebCore::originatingStyleableIncludingTimelineScope):
(WebCore::originatingElementExcludingTimelineScope):
(WebCore::AnimationTimelinesController::relatedTimelineScopeElements):
(WebCore::containsRenderer):
(WebCore::determineTreeOrder):
(WebCore::determineTimelineForElement):
(WebCore::AnimationTimelinesController::updateTimelineForTimelineScope):
(WebCore::AnimationTimelinesController::registerNamedScrollTimeline):
(WebCore::AnimationTimelinesController::attachPendingOperations):
(WebCore::AnimationTimelinesController::registerNamedViewTimeline):
(WebCore::AnimationTimelinesController::unregisterNamedTimeline):
(WebCore::AnimationTimelinesController::setTimelineForName):
(WebCore::updateTimelinesForTimelineScope):
(WebCore::AnimationTimelinesController::updateNamedTimelineMapForTimelineScope):
(WebCore::AnimationTimelinesController::unregisterNamedTimelinesAssociatedWithElement):
(WebCore::originatingElementIncludingTimelineScope): Deleted.
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncStyleOriginatedTimeline):
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::setTimelineScopeStyleable):
(WebCore::ScrollTimeline::setTimelineScopeElement): Deleted.
* Source/WebCore/animation/ScrollTimeline.h:
(WebCore::ScrollTimeline::sourceStyleable const):
(WebCore::ScrollTimeline::timelineScopeDeclaredStyleable const):
(WebCore::ScrollTimeline::clearTimelineScopeDeclaredStyleable):
(WebCore::ScrollTimeline::timelineScopeDeclaredElement const): Deleted.
(WebCore::ScrollTimeline::clearTimelineScopeDeclaredElement): Deleted.
* Source/WebCore/animation/ViewTimeline.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::cancelStyleOriginatedAnimations const):
(WebCore::Styleable::updateCSSScrollTimelines const):
(WebCore::Styleable::updateCSSViewTimelines const):

Canonical link: <a href="https://commits.webkit.org/289913@main">https://commits.webkit.org/289913@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc419ee13bc8a3bd83b57638222f42706b348b20

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88355 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7874 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42781 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93312 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39109 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90406 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8261 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16057 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68149 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25869 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91357 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6314 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79911 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48518 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6086 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34322 "Found 4 new test failures: accessibility/negative-tabindex-does-not-expose-label.html imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/block-container-block-start-self-collapsing-nested.html imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/track/track-element/track-cues-cuechange.html imported/w3c/web-platform-tests/html/semantics/embedded-content/the-iframe-element/iframe-loading-lazy-multiple-queued-navigations.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38217 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76468 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35206 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95155 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15530 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11380 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77015 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15786 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75764 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76265 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20659 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19075 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8567 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13815 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15546 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/20853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15287 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18736 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17069 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/pseudo-on-scroller-named-timeline.html:
```diff
@@ -0,0 +1,64 @@
+  width: 50px;
+  height: 50px;
+  background: red;
+  animation: bg linear reverse;
```
Is `reverse` here required for this test?

nmoucht replied:
Not strictly no.

graouts commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -257,9 +257,9 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ScrollTimeline::documentW
+void ScrollTimeline::setTimelineScopeStyleable(const Styleable& styleable)
 {
-    m_timelineScopeElement = WeakPtr { &element };
+    m_timelineScopeStyleable = styleable;
```
It does not appear that there is testing coverage for `timeline-scope` being set on a pseudo-element, which this patch now allows.

nmoucht replied:
Yeah I think this should be factored out and made a separate patch.

graouts reviewed (approved):
You need to export the WPT test.

nmoucht committed:
[scroll-animations] Have named timeline code use Styleable
https://bugs.webkit.org/show_bug.cgi?id=286781
rdar://143926206

Reviewed by Antoine Quint.

Have named timeline code use Styleable.

* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::originatingElement):
(WebCore::originatingStyleableIncludingTimelineScope):
(WebCore::originatingElementExcludingTimelineScope):
(WebCore::AnimationTimelinesController::relatedTimelineScopeElements):
(WebCore::containsRenderer):
(WebCore::determineTreeOrder):
(WebCore::determineTimelineForElement):
(WebCore::AnimationTimelinesController::updateTimelineForTimelineScope):
(WebCore::AnimationTimelinesController::registerNamedScrollTimeline):
(WebCore::AnimationTimelinesController::attachPendingOperations):
(WebCore::AnimationTimelinesController::registerNamedViewTimeline):
(WebCore::AnimationTimelinesController::unregisterNamedTimeline):
(WebCore::AnimationTimelinesController::setTimelineForName):
(WebCore::updateTimelinesForTimelineScope):
(WebCore::AnimationTimelinesController::updateNamedTimelineMapForTimelineScope):
(WebCore::AnimationTimelinesController::unregisterNamedTimelinesAssociatedWithElement):
(WebCore::originatingElementIncludingTimelineScope): Deleted.
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncStyleOriginatedTimeline):
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::setTimelineScopeStyleable):
(WebCore::ScrollTimeline::setTimelineScopeElement): Deleted.
* Source/WebCore/animation/ScrollTimeline.h:
(WebCore::ScrollTimeline::sourceStyleable const):
(WebCore::ScrollTimeline::timelineScopeDeclaredStyleable const):
(WebCore::ScrollTimeline::clearTimelineScopeDeclaredStyleable):
(WebCore::ScrollTimeline::timelineScopeDeclaredElement const): Deleted.
(WebCore::ScrollTimeline::clearTimelineScopeDeclaredElement): Deleted.
* Source/WebCore/animation/ViewTimeline.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::cancelStyleOriginatedAnimations const):
(WebCore::Styleable::updateCSSScrollTimelines const):
(WebCore::Styleable::updateCSSViewTimelines const):

Canonical link: https://commits.webkit.org/289913@main
